src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,411 @@
+ * lists (CRLs).  It is defined in RFC 1421 and RFC 7468.  PEM consists of a
+ * Base64-formatted binary encoding enclosed by a type-identifying header
+ * and footer.
+ *
+ * <p> Decoding methods return an instance of a class that matches the data
+ * type and implements {@link DEREncodable} unless otherwise specified. The
+ * following types are decoded into Java Cryptographic Extensions (JCE) object
+ * representations:
+ * <pre>
+ *     PRIVATE KEY, RSA PRIVATE KEY, PUBLIC KEY, CERTIFICATE,
+ *     X509 CERTIFICATE, X509 CRL, and ENCRYPTED PRIVATE KEY.
+ * </pre>
+ *
+ * A specified return class must implement {@link DEREncodable} and be an
+ * appropriate JCE object class for the PEM; otherwise an
+ * {@link IllegalArgumentException} is thrown.
+ *
+ * <p> If the PEM does not have a JCE object representation, it returns a
+ * {@link PEMRecord}. Any PEM can be decoded into a {@code PEMRecord} if the
+ * class is specified.
+ *
+ * <p> A new immutable {@code PEMDecoder} instance is created when configured
+ * with {@linkplain #withFactory} and/or {@linkplain #withDecryption}.
+ * Configuring an instance for decryption does not prevent decoding with
+ * unencrypted PEM. Any encrypted PEM that does not use the configured password
+ * will throw an {@link SecurityException}.  A decoder instance not configured
+ * with decryption returns a {@link EncryptedPrivateKeyInfo} with encrypted
+ * PEM.  {@code EncryptedPrivateKeyInfo} methods must be used to retrieve the
+ * {@link PrivateKey}.
+ *
+ * <p> Byte streams consumed by methods in this class are assumed to represent
+ * characters encoded in the
+ * {@link java.nio.charset.StandardCharsets#ISO_8859_1 ISO-8859-1} charset.
+ *
+ * @apiNote
+ * Here is an example of decoding a PrivateKey object:
+ * <pre>
+ *     PEMDecoder pd = PEMDecoder.of();
+ *     PrivateKey priKey = pd.decode(priKeyPEM);
+ * </pre>
+ *
+ * @spec https://www.rfc-editor.org/info/rfc1421
+ *       RFC 1421: Privacy Enhancement for Internet Electronic Mail
+ * @spec https://www.rfc-editor.org/info/rfc7468
+ *       RFC 7468: Textual Encodings of PKIX, PKCS, and CMS Structures
+ *
+ * @since 25
+ */
+
+@PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
+public final class PEMDecoder {
+    private final Provider factory;
+    private final PBEKeySpec password;
+
+    // Singleton instance for PEMDecoder
+    private final static PEMDecoder PEM_DECODER = new PEMDecoder(null, null);
+
+    /**
+     * Creates an immutable instance with a specific KeyFactory and/or
+     * password.
+     * @param withFactory KeyFactory provider
+     * @param withPassword char[] password for EncryptedPrivateKeyInfo
+     *                    decryption
+     */
+    private PEMDecoder(Provider withFactory, PBEKeySpec withPassword) {
+        password = withPassword;
+        factory = withFactory;
+    }
+
+    /**
+     * Returns an instance of {@code PEMDecoder}.  This instance may be repeatedly used
+     * to decode different PEM text.
```

Comment:
Not only repeatable, but also thread safe I assume? Maybe we need to talk about this in the class spec. Immutable does not automatic implies thread-safe, if the immutability is only observed on the API level.

Reply:
Usually there is a sentence at the end of the class description, ex- "This class is immutable and thread-safe." Once you state that, you don't need to say anything more in the methods like the sentence above about repeatability.